src/pudl/transform/ferc1.py

```diff
@@ -891,6 +883,68 @@ def reconcile_table_calculations(
     return calculated_df
 
 
+def calculate_values_from_components(
+    calculation_components: pd.DataFrame,
+    data: pd.DataFrame,
+    validate: Literal["one_to_many", "many_to_many"],
+    calc_idx: list[str],
+    data_idx: list[str],
+    value_col: str,
+) -> pd.DataFrame:
+    """Calculate values from components.
+
+    Add a ``calculated_amount`` column to the ``data`` table for all calculated values
+    using the componentents within ``calculation_components``.
+    """
+    # Merge the reported data and the calculation component metadata to enable
+    # validation of calculated values. Here the data table exploded is supplying the
+    # values associated with individual calculation components, and the table_name
+    # and xbrl_factoid to which we aggregate are coming from the calculation
+    # components table. After merging we use the weights to adjust the reported
+    # values so they can be summed directly. This gives us aggregated calculated
+    # values that can later be compared to the higher level reported values.
+
+    # we are going to merge the data onto the calc components with the _parent
+    # column names, so the groupby after the merge needs a set of by cols with the
+    # _parent suffix
+    gby_parent = [
+        f"{col}_parent" if col in ["table_name", "xbrl_factoid"] else col
+        for col in data_idx
+    ]
```

Comment:
Sorry if we talked about this yesterday and I forgot but why don't we need to group by all of the parent key columns, including the dimensions? Why is it only using table name and factoid?

Reply:
Ah, I think I was maybe I was thrown off by this conditional comprehension.  I don't understand what you expect to have in here now. It seems like a mix of `_parent` and non-parent columns. Why is that appropriate? Is there a way we can make this more readable?

Reply:
the only columns that have the `_parent` suffix is the "table_name" and "xbrl_factoid"

Reply:
we didn't talk about this yesterday but we did talk about this for a while in my last pr https://github.com/catalyst-cooperative/pudl/pull/2753!

if it would be more clear, I could check for any _parent suffixed columns in calculation_components and replace the `["table_name", "xbrl_factoid"]` with that list.